dbms/src/Flash/ResourceControl/LocalAdmissionController.h

```diff
@@ -158,22 +336,245 @@ class ResourceGroup final : private boost::noncopyable
+
     double ru_consumption_delta = 0.0;
 
     LoggerPtr log;
+
+    std::chrono::time_point<std::chrono::steady_clock> last_fetch_tokens_from_gac_timepoint
+        = std::chrono::steady_clock::now();
+    std::chrono::time_point<std::chrono::steady_clock> stop_trickle_timepoint = std::chrono::steady_clock::now();
 };
 
 using ResourceGroupPtr = std::shared_ptr<ResourceGroup>;
 
+// LocalAdmissionController is the local(tiflash) part of the distributed token bucket algorithm.
+// It manages all resource groups:
+// 1. Creation, deletion and config updates of resource group.
+// 2. Fetching tokens from GAC periodically or when tokens are low.
+// 3. Record/report resource consumption and the priority of each resource group.
 class LocalAdmissionController final : private boost::noncopyable
 {
 public:
+    // For tidb_enable_resource_control is disabled.
+    static constexpr uint64_t HIGHEST_RESOURCE_GROUP_PRIORITY = 0;
+
+    LocalAdmissionController(::pingcap::kv::Cluster * cluster_, Etcd::ClientPtr etcd_client_)
+        : cluster(cluster_)
+        , etcd_client(etcd_client_)
+        , thread_manager(newThreadManager())
+    {
+        thread_manager->scheduleThenDetach(true, "LocalAdmissionController::fetchGAC", [this] {
+            this->startBackgroudJob();
+        });
+        thread_manager->scheduleThenDetach(true, "LocalAdmissionController::watchGAC", [this] { this->watchGAC(); });
+    }
+
+    ~LocalAdmissionController() { stop(); }
+
+    // NOTE: getOrFetchResourceGroup may throw if resource group has been deleted.
+    void consumeResource(const std::string & name, double ru, uint64_t cpu_time_in_ns)
+    {
+        // When tidb_enable_resource_control is disabled, resource group name is empty.
+        if (name.empty())
+            return;
+
+        ResourceGroupPtr group = getOrFetchResourceGroup(name);
+        group->consumeResource(ru, cpu_time_in_ns);
+        if (group->lowToken())
+        {
+            {
+                std::lock_guard lock(mu);
+                low_token_resource_groups.insert(name);
+            }
+            cv.notify_one();
+        }
+    }
+
+    uint64_t getPriority(const std::string & name)
+    {
+        if (name.empty())
+            return HIGHEST_RESOURCE_GROUP_PRIORITY;
+
+        ResourceGroupPtr group = getOrFetchResourceGroup(name);
+        return group->getPriority(max_ru_per_sec.load());
+    }
+
+    bool isResourceGroupThrottled(const std::string & name)
+    {
+        if (name.empty())
+            return false;
+
+        // todo: refine: ResourceControlQueue can record the statics of running pipeline task.
+        // If the proportion of scheduled pipeline task is less than 10%, we say this resource group is throttled.
+        ResourceGroupPtr group = findResourceGroup(name);
+        if (group == nullptr)
+            return false;
+        return group->getRU() <= 0.0;
+    }
+
+    std::string isResourceGroupValid(const std::string & name)
```

Comment:
Where is this function used?
And it's a bit strange that isXXX returns a string.

Reply:
`isResourceGroupThrottled` and `isResourceGroupValid` removed. It's used for the former code of min tso.